src/understand-tidb/privilege-management.md

```diff
@@ -1,2 +1,58 @@
+
+Because the asynchronous notifications do not guarantee delivery, TiDB will also [automatically rebuild the privilege cache](https://github.com/pingcap/tidb/blob/5e05922de6a253859cfbfe19356de8a2e2db39da/domain/domain.go#L852-L908) every 5-10 minutes in a loop. This behavior is not strictly MySQL compatible, because in MySQL the privilege cache will only ever be rebuilt from a `FLUSH PRIVILEGES` statement, a restart, or a privilege control statement.
+
+## Adding privilege checks to a statement
+
+Some privilege checks are automatically assigned during plan building, for example ensuring that you have permissions to the tables that will be accessed. These checks are skipped for `information_schema` tables, and should you add an additional statement (such as `SHOW xyz`), you will also need to ensure that privilege checks are added.
+
+Should you need to add privilege checks there are two options:
+
+1. During plan building you can attach `visitInfo` to the plan (examples: [`SET CONFIG`](https://github.com/pingcap/tidb/blob/5e05922de6a253859cfbfe19356de8a2e2db39da/planner/core/planbuilder.go#L745), [`SHOW BACKUPS`](https://github.com/pingcap/tidb/blob/5e05922de6a253859cfbfe19356de8a2e2db39da/planner/core/planbuilder.go#L2378-L2380))
+
+2. In the executor function which handles the statement (examples: [`DROP USER`](https://github.com/pingcap/tidb/blob/5e05922de6a253859cfbfe19356de8a2e2db39da/executor/simple.go#L1150-L1170)).
+
+The first option is recommended, as it is much less verbose. However, `visitInfo` can only attach additional privilege checks with an `AND` semantic. In the `DROP USER` example linked above, the MySQL compatible behavior is that either the `DELETE` privilege on the `mysql.user` table *or* the `CREATE USER` privilege is required (with some additional exceptions on if the user is a "SYSTEM_USER").
+
+There are also similar complex examples in `information_schema` tables and commands such as `SHOW PROCESSLIST` where the result of the statement depends on the permissions of the user executing it.
+
+### Manually checking with the privilege manager
+
+For (2) above, manual checks should follow the following pattern:
+```
+checker := privilege.GetPrivilegeManager(e.ctx)
+if checker != nil && !checker.RequestVerification(ctx.GetSessionVars().ActiveRoles, schema.Name.L, table.Name.L, "", mysql.AllPrivMask) {
+    /* do something */
+}
+..
+if checker == nil || !checker.RequestDynamicVerification(ctx.GetSessionVars().ActiveRoles, "RESTRICTED_TABLES_ADMIN", false) {
+    /* do something */
+}
+```
+
+The check for `checker != nil` is important because for internal SQL statements the privilege manager is not present. These statements are expected to fall through and satisfy the privilege check.
```

Comment:
Although I can get the point you want to express, the example above of DROP USER actually fails on `check == nil`

```go
	checker := privilege.GetPrivilegeManager(e.ctx)
	if checker == nil {
		return errors.New("miss privilege checker")
	}
```

may we choose a better example or elaborate a bit here?

Reply:
Yeah, I will chose a better example :-)